### ISSUE TITLE
Warning on needless location-specific disable directives

### ISSUE THREAD

**Issue by orenbenkiki** [PyCQA/pylint#3796]
**Example**

When writing:

```python
def foo():  # pylint: disable=too-many-statements
    return 'bar'
```

Have the option of `pylint` generate a warning that there were not too-many-statements in the 1st place.

This is only relevant to directives that apply to a specific location, as opposed to directives that apply to a whole code region. That is, the following would *not* trigger the warning:

```python
# pylint: disable=too-many-statements

def foo(): 
    return 'bar'
```

**Motivation**

For me, `pylint` serves a dual purpose. First, it does help me keep any new code clean and readable. But in addition, the `disable` directives serve as visible reminders of code smells directly in the source.

To achieve this, I ensure `pylint` doesn't emit any warnings on my code, by adding specific `disable` directives where the issues isn't fixed (yet). This might be accompanied by a comment justifying not fixing the issue, but in most cases, the issue should be resolved, and either way, it helps the reader by warning him that the following code is tricker than it should be.

However, this usage model breaks when some refactoring removes the original issue, but the `disable` directive remains.

I use the same model for `mypy` type annotations. There, `mypy` will automatically warn you about any needless `type: ignore` directives, which means that when I see one I know for sure there's something tricky going on that `mypy` can't figure out (and needs to be clarified by the code itself or by a comment).

**Alternatives**

Naively, this should be easier to implement than #3680 and give the same benefits to the reader (the confidence the disabled issue actually exists). It wouldn't be as convenient to the code author, who will need to manually add each directive, but that is arguably a good thing - a bit more motivation for fixing the issue rather than just slapping a `disable` directive on it.

**Comment by dbaty** [comment]
Are you aware of the `useless-suppression` checker? In your examples, it would warn about both "# pylint: disable=..." comments. I understand that you would not want to be warned about the module-level comment. Still, I think it's a good compromise.

There is a catch, though: this checker yields an "informational message", not a warning. As such, pylint exits with a success code. If you pylint to fail, you can pip-install and enable the `pylint-strict-informational` package, which will turn "informational messages" into warnings. You would do that like this in your `.pylintrc`:

```ini
[MASTER]
load-plugins=
    pylint_strict_informational


[MESSAGES CONTROL]
enable=
    useless-suppression
```

We do exactly that on our company projects. It works well: when we enabled it, it showed us quite a large number of useless ignore directives that we had accumulated for years! :)

Hope it helps!

**Comment by orenbenkiki** [comment]
Somehow I missed that in my google searched...

One point though, it seems `useless-suppressions` implies `suppressed-message`, so if one enables the former one should also disable the latter (otherwise you get a message for each `disable` directive in the code).

Many thanks!

BTW - only somewhat question - is it possible to have pylint emit the score of the code disregarding suppressions? That is, the score it would have given the code if all the suppressions were removed?

**Comment by dbaty** [comment]
> One point though, it seems useless-suppressions implies suppressed-message, so if one enables the former one should also disable the latter (otherwise you get a message for each disable directive in the code).

As far as I understand, it's not necessary if `suppressed-message` is enabled in your configuration file.

> BTW - only somewhat question - is it possible to have pylint emit the score of the code disregarding suppressions? That is, the score it would have given the code if all the suppressions were removed?

I don't think that's currently possible (and, not knowing enough about pylint internals, I don't know how it could be done).

**Comment by orenbenkiki** [comment]
> > One point though, it seems useless-suppressions implies suppressed-message, so if one enables the former one should also disable the latter (otherwise you get a message for each disable directive in the code).
> 
> As far as I understand, it's not necessary if `suppressed-message` is enabled in your configuration file.

Yes, sorry, I just checked again and it was due to a typo.

As for the score, it might have been nice to distinguish between two "intents" of suppressing a message: `disable` as in "I know what I'm doing, it is OK" and `defer` as in "You are right, but don't complain about it, I have marked this to be fixed later" - where the latter suppressed messages would still count against the score.

Not that this is a crucial feature...
